src/libproc/proc_pid.rs

```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
+///                                     let mut addr = 0;
+///                                     addr |= s_addr >> 24 & 0x000000ff;
+///                                     addr |= s_addr >> 8  & 0x0000ff00;
+///                                     addr |= s_addr << 8  & 0x00ff0000;
+///                                     addr |= s_addr << 24 & 0xff000000;
+///
+///                                     println!("{}.{}.{}.{}:{}", addr >> 24 & 0xff, addr >> 16 & 0xff, addr >> 8 & 0xff, addr & 0xff, port);
+///                                 },
+///                                 _ => (),
+///                             }
+///                         }
+///                     },
+///                     _ => (),
+///                 }
+///             }
+///         }
+///     }
+/// }
+/// ```
+///
+pub fn pidfdinfo<T: PIDFDInfo>(pid : i32, fd: int32_t) -> Result<T, String> {
+    let flavor = T::flavor() as i32;
+    let buffer_size = mem::size_of::<T>() as i32;
+    let mut pidinfo = T::default();
+    let buffer_ptr = &mut pidinfo as *mut _ as *mut c_void;
+    let ret: i32;
+
+    unsafe {
+        ret = proc_pidfdinfo(pid, fd, flavor, buffer_ptr, buffer_size);
+    };
+
+    if ret <= 0 {
+        Err(get_errno_with_message(ret))
+    } else {
+        Ok(pidinfo)
+    }
+}
+
+#[test]
+fn pidfdinfo_test() {
+    use std::process;
+    use std::net::TcpListener;
+    let pid = process::id() as i32;
+
+    let _listener = TcpListener::bind("127.0.0.1:65535");
+
+    match pidinfo::<BSDInfo>(pid, 0) {
+        Ok(info) => {
+            match listpidinfo::<ListFDs>(pid, info.pbi_nfiles as usize) {
+                Ok(fds) => {
+                    for fd in fds {
+                        match fd.proc_fdtype.into() {
+                            ProcFDType::Socket => {
+                                if let Ok(socket) = pidfdinfo::<SocketFDInfo>(pid, fd.proc_fd) {
+                                    match socket.psi.soi_kind.into() {
+                                        SocketInfoKind::Tcp => unsafe {
+                                            let info = socket.psi.soi_proto.pri_tcp;
+                                            assert_eq!(socket.psi.soi_protocol,
+                                            libc::IPPROTO_TCP);
+                                            assert_eq!(info.tcpsi_ini.insi_lport as u32, 65535);
+                                        }
+                                        _ => (),
+                                    }
+                                }
+                            },
+                            _ => (),
+                        }
+                    }
+                },
+                Err(err) => assert!(false, "Error retrieving process info: {}", err)
+            }
+        },
+        Err(err) => assert!(false, "Error retrieving process info: {}", err)
+    };
+}
+
+#[repr(C)]
+#[derive(Default)]
```

Comment:
Do you actually need Default on all of these structs?

Reply:
This Default is used in `pidfdinfo` like `let mut pidinfo = T::default();`.
I think all members of `SocketFDInfo` must have Default.